### PR TITLE
image management

### PR DESCRIPTION
**Pull request by bcumming**
An omnibus PR that introduces a raft of features:

- support for managing images using the `uenv image` sub-command
- much improved CLI help commands
- move common code to a library
- add some basic unit tests for the python code. These can be expanded as we fix issues.

### CODE REVIEW

File: lib/alps.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# arbitrary alps-specific gubbins
+
+import os
```
Comment (simonpintarelli) [inline lib/alps.py:3]: `import terminal` is missing here.

File: uenv-image [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+    elif args.command == "inspect":
+        terminal.info(f"repo path: {repo_path}")
+
+        img_filter = get_filter(args)
+
+        try:
+            fscache = datastore.FileSystemRepo(repo_path)
+        except FileNotFoundError as err:
+            terminal.error(f"""the local repository {repo_path} does not exist.
+If the repo does not exist, you can try creating it first: {terminal.colorize(f"uenv image create {repo_path}", "white")}
```
Comment (simonpintarelli) [inline uenv-image]: I think the `-r` flag is missing to `image create`. Or should it just be `uenv image create`?

**Comment by simonpintarelli** [comment]
Would be nice  to have a meaningful error message for `uenv image pull` too (same as for `uenv image inspect`) in case the repository has not yet been created.
```
[santis][simonpi@santis-ln001 simonpi]$ uenv image pull icon-wcp/v1:latest  
Traceback (most recent call last):
  File "/users/simonpi/.local/libexec/uenv-image", line 399, in <module>
    cache = datastore.FileSystemRepo(repo_path)
  File "/users/simonpi/.local/libexec/lib/datastore.py", line 269, in __init__
    raise FileNotFoundError(f"uenv image repository not found {self._path}: the path does not exist.")
FileNotFoundError: uenv image repository not found /bret/scratch/cscs/simonpi/.uenv-images: the path does not exist.
```

File: uenv-image [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+        except names.IncompleteUenvName as err:
+            terminal.error(f"source {source} is not fully qualified: use name/version:tag or sha256.")
+
+        # query JFrog for the list of images
+        try:
+            _, build_database = jfrog.query()
+        except RuntimeError as err:
+            terminal.error(f"{str(err)}")
+
+        terminal.info(f"downloaded jfrog build meta data: {len(build_database.images)} images")
```
Comment (simonpintarelli) [inline uenv-image]: `build_database.images` has type `RecordSet`, but it has no `__len__` method.

**Comment by bcumming** [comment]
> Would be nice to have a meaningful error message for `uenv image pull` too (same as for `uenv image inspect`) in case the repository has not yet been created.
> 
> ```
> [santis][simonpi@santis-ln001 simonpi]$ uenv image pull icon-wcp/v1:latest  
> Traceback (most recent call last):
>   File "/users/simonpi/.local/libexec/uenv-image", line 399, in <module>
>     cache = datastore.FileSystemRepo(repo_path)
>   File "/users/simonpi/.local/libexec/lib/datastore.py", line 269, in __init__
>     raise FileNotFoundError(f"uenv image repository not found {self._path}: the path does not exist.")
> FileNotFoundError: uenv image repository not found /bret/scratch/cscs/simonpi/.uenv-images: the path does not exist.
> ```

Agreed. I will add a function that catches this error and generates a nice user-friendly error message.

**Review by simonpintarelli — APPROVED** [review]
lgtm

**Pull request merged**